ansible-galaxy: forgive meta/main.yml mishaps

willthames opened the pull request:
##### Issue Type:

Bugfix Pull Request
##### Ansible Version:

[wthames@wthames ansible (devel)]$ ansible --version
ansible 1.9 (devel c8de03d516) last updated 2015/03/13 11:14:59 (GMT +1000)
  lib/ansible/modules/core: (detached HEAD 31cc5f543f) last updated 2015/03/13 11:15:44 (GMT +1000)
  lib/ansible/modules/extras: (detached HEAD 8baba98ebe) last updated 2015/03/13 11:15:06 (GMT +1000)
  v2/ansible/modules/core: (detached HEAD 34784b7a61) last updated 2015/03/05 14:11:54 (GMT +1000)
  v2/ansible/modules/extras: (detached HEAD 46e316a20a) last updated 2015/03/05 14:11:54 (GMT +1000)
  configured module search path = None
##### Environment:

N/A
##### Summary:

For roles not created through ansible-galaxy init, but
manually, meta/main.yml may not exist or contain
dependency information. Handle such cases gracefully
- warn but cope when meta/main.yml is missing
- warn but cope when meta/main.yml is inaccessible or unparseable
- warn but cope when meta/main.yml contains no dependencies data
- warn but cope when meta/main.yml effectively has dependencies: None

The latter case looks like:

```
dependencies:
```

All of the above cases are treated as an empty dependency list.
##### Steps To Reproduce:

Create a role without a meta/main.yml file and try to install it
Create a role with a meta/main.yml but with no dependencies section and try to install it

Cope with other general problems like inaccessible or unparseable meta/main.yml or a null dependencies section
##### Expected Results:

A warning is generated but role installation continues
##### Actual Results:

Currently ansible-galaxy install bombs out when meta/main.yml is missing or broken with a NoneType error

```
Traceback (most recent call last):
  File "/home/wthames/src/ansible/bin/ansible-galaxy", line 973, in <module>
    main()
  File "/home/wthames/src/ansible/bin/ansible-galaxy", line 967, in main
    fn(args, options, parser)
  File "/home/wthames/src/ansible/bin/ansible-galaxy", line 854, in execute_install
    for dep in role_dependencies:
TypeError: 'NoneType' object is not iterable
```

willthames committed:
ansible-galaxy: forgive meta/main.yml mishaps

For roles not created through ansible-galaxy init, but
manually, meta/main.yml may not exist or contain
dependency information. Handle such cases gracefully

* warn but cope when meta/main.yml is missing
* warn but cope when meta/main.yml is inaccessible or unparseable
* warn but cope when meta/main.yml contains no dependencies data
* warn but cope when meta/main.yml effectively has dependencies: None

The latter case looks like:
```
dependencies:
```

All of the above cases are treated as an empty dependency list.

bcoca commented on bin/ansible-galaxy:
```diff
@@ -414,9 +414,11 @@ def get_role_metadata(role_name, options):
             return None
     except:
-        return None    
+        print '- WARN: cannot access or parse role %s meta/main.yml' % role_name
```
use the utils warning function instead of a print

bcoca commented on bin/ansible-galaxy:
```diff
@@ -414,9 +414,11 @@ def get_role_metadata(role_name, options):
             f.close()
             return meta_data
         else:
+            print '- WARN: role %s does not contain a meta/main.yml' % role_name
```
too noisy, no need to warn in this case

bcoca commented on bin/ansible-galaxy:
```diff
@@ -842,9 +844,16 @@ def execute_install(args, options, parser):
+                        role_dependencies = []
+                    else:
+                        if role_data.get('dependencies') is None:
+                            print '- WARN: %s meta/main.yml does not contain a dependencies list' % role.get('name')
```
same here, no dependencies is OK

willthames commented:
Ok, did some more testing based on your suggested changes and can actually remove a lot of cruft if you don't need a meta/main.yml to install a role. I've left the check for meta/main.yml when force installing a role for now (i.e. remove_role won't remove a role without a meta/main.yml)

I've moved to ansible.utils.warnings for warnings - that now makes them more dramatic than the existing error messages.

willthames committed:
ansible-galaxy: remove unnecessary warnings

Moved to ansible.utils.warnings for warnings
Remove unnecessary meta/main.yml checks from install_role
Improve error messages

bcoca commented on bin/ansible-galaxy:
```diff
@@ -513,20 +514,6 @@ def install_role(role_name, role_version, role_filename, options):
-                meta_file_data = yaml.safe_load(role_tar_file.extractfile(meta_file))
-            except:
-                print "- error: this role does not appear to have a valid meta/main.yml file."
-                return False
```
instead of removing these, wouldn't it make sense to make them warnings?

willthames replied:
We already warn for unparseable/unreadable meta/main.yml in `get_role_metadata`.

I only removed the remaining messages following your earlier comments on missing meta/main.yml

> too noisy, no need to warn in this case

amenonsen commented:
@willthames do you expect to be able to rebase this and fix the failing tests and resubmit?

willthames commented:
@amenonsen, it's really down to @bcoca. There's no point me revisiting this pull request until galaxy settles down.

bcoca commented:
galaxy should be stable enough now to attempt to redo this

willthames commented:
@bcoca, my meaning of stable is that `make test_galaxy` executes the existing tests, and does not fail. Let me know when that is the case.

bcoca commented:
As far as i can tell these tests were never run successfully nor added to the integration suite, I just tested with 1.9 and they fail. Was this ever not the case?

willthames commented:
@bcoca, tests work fine in stable-1.9 branch if you have hg and git installed (if not, let me know, perhaps the repos in question aren't set up right)

I used these tests as part of the validation of all the 1.8 improvements, and they should continue to be used to validate 2.0 even implements the features of the 1.8 release.

willthames commented:
You also need to be able to git clone using git@github.com (so basically if you have push access to any github repo (e.g. your own), the tests should work)

willthames commented:
```
$ make test_galaxy
mytmpdir=/tmp/tmp.nCgJ78E74X ; \
ansible-galaxy install -r galaxy_rolesfile -p $mytmpdir/roles ; \
    cp galaxy_playbook.yml $mytmpdir ; \
    ansible-playbook -i inventory $mytmpdir/galaxy_playbook.yml -v  ; \
    RC=$? ; \
    rm -rf $mytmpdir ; \
    exit $RC
- executing: git clone http://bitbucket.org/willthames/git-ansible-galaxy git-ansible-galaxy
- executing: git archive --prefix=git-ansible-galaxy/ --output=/tmp/tmpf38BeY.tar v1.4
- extracting git-ansible-galaxy to /tmp/tmp.nCgJ78E74X/roles/git-ansible-galaxy
- git-ansible-galaxy was installed successfully
- adding dependency: git-gal-dep
- executing: hg clone http://bitbucket.org/willthames/hg-ansible-galaxy hg-ansible-galaxy
- executing: hg archive --prefix hg-ansible-galaxy/ -r tip /tmp/tmplFAm3t.tar
- extracting hg-ansible-galaxy to /tmp/tmp.nCgJ78E74X/roles/hg-ansible-galaxy
- hg-ansible-galaxy was installed successfully
- downloading role from https://bitbucket.org/willthames/http-ansible-galaxy/get/master.tar.gz
- extracting http-role to /tmp/tmp.nCgJ78E74X/roles/http-role
- http-role was installed successfully
- executing: git clone git@github.com:geerlingguy/ansible-role-php.git ansible-role-php
- executing: git archive --prefix=ansible-role-php/ --output=/tmp/tmpda2W4H.tar master
- extracting ansible-role-php to /tmp/tmp.nCgJ78E74X/roles/ansible-role-php
- ansible-role-php was installed successfully
- executing: git clone https://bitbucket.org/willthames/git-ansible-galaxy-dep git-gal-dep
- executing: git archive --prefix=git-gal-dep/ --output=/tmp/tmpUIm0Z4.tar v0.1
- extracting git-gal-dep to /tmp/tmp.nCgJ78E74X/roles/git-gal-dep
- git-gal-dep was installed successfully

PLAY [localhost] ************************************************************** 

GATHERING FACTS *************************************************************** 
ok: [localhost]

TASK: [git-gal-dep | dependency check] **************************************** 
ok: [localhost] => {
    "msg": "Hello world!"
}

TASK: [git-ansible-galaxy | double check] ************************************* 
ok: [localhost] => {
    "msg": "Hello!"
}

TASK: [http-role | check this runs] ******************************************* 
ok: [localhost] => {
    "msg": "Downloading from a tar file seems to work"
}

PLAY RECAP ******************************************************************** 
localhost                  : ok=4    changed=0    unreachable=0    failed=0   

mytmpdir=/tmp/tmp.WORa7qqoZ4 ; \
ansible-galaxy install -r galaxy_roles.yml -p $mytmpdir/roles ; \
    cp galaxy_playbook.yml $mytmpdir ; \
    ansible-playbook -i inventory $mytmpdir/galaxy_playbook.yml -v  ; \
    RC=$? ; \
    rm -rf $mytmpdir ; \
    exit $RC
- executing: git clone http://bitbucket.org/willthames/git-ansible-galaxy git-ansible-galaxy
- executing: git archive --prefix=git-ansible-galaxy/ --output=/tmp/tmptUioqy.tar v1.4
- extracting git-ansible-galaxy to /tmp/tmp.WORa7qqoZ4/roles/git-ansible-galaxy
- git-ansible-galaxy was installed successfully
- adding dependency: git-gal-dep
- executing: hg clone http://bitbucket.org/willthames/hg-ansible-galaxy hg-ansible-galaxy
- executing: hg archive --prefix hg-ansible-galaxy/ /tmp/tmpOnOulP.tar
- extracting hg-ansible-galaxy to /tmp/tmp.WORa7qqoZ4/roles/hg-ansible-galaxy
- hg-ansible-galaxy was installed successfully
- downloading role from https://bitbucket.org/willthames/http-ansible-galaxy/get/master.tar.gz
- extracting http-role to /tmp/tmp.WORa7qqoZ4/roles/http-role
- http-role was installed successfully
- executing: git clone git@github.com:geerlingguy/ansible-role-php.git php
- executing: git archive --prefix=php/ --output=/tmp/tmpyNpZvc.tar HEAD
- extracting php to /tmp/tmp.WORa7qqoZ4/roles/php
- php was installed successfully
- executing: git clone https://bitbucket.org/willthames/git-ansible-galaxy-dep git-gal-dep
- executing: git archive --prefix=git-gal-dep/ --output=/tmp/tmpQlY8XP.tar v0.1
- extracting git-gal-dep to /tmp/tmp.WORa7qqoZ4/roles/git-gal-dep
- git-gal-dep was installed successfully

PLAY [localhost] ************************************************************** 

GATHERING FACTS *************************************************************** 
ok: [localhost]

TASK: [git-gal-dep | dependency check] **************************************** 
ok: [localhost] => {
    "msg": "Hello world!"
}

TASK: [git-ansible-galaxy | double check] ************************************* 
ok: [localhost] => {
    "msg": "Hello!"
}

TASK: [http-role | check this runs] ******************************************* 
ok: [localhost] => {
    "msg": "Downloading from a tar file seems to work"
}

PLAY RECAP ******************************************************************** 
localhost                  : ok=4    changed=0    unreachable=0    failed=0   
```

bcoca commented:
I have mercurial, but several of the repos fail to clone, both git and hg uris. I"m guessing this failed on our internal tests machines also and is why they were never added to the normal test suite.

bcoca commented:
ansible-role-php and git-gal-dep seem to be the problems

willthames commented:
@bcoca what happens when you try and clone those URLs manually? 

The git@github.com might cause problems on an integration test machine if no ssh key is available, but the rest should be ok. It's useful to have the ssh test case (because that URL should work too, and indeed we rely on it to provide access to roles in private repos). 

This test suite is definitely worth having, I can't see why the issue of failing tests wasn't raised earlier (and I've raised my concerns that make test_galaxy clearly hasn't been run a number of times now)

willthames commented:
https://github.com/ansible/ansible/search?q=test_galaxy&type=Issues&utf8=%E2%9C%93

You even closed one of them as a dupe.

bcoca commented:
We probably need to put these in destructive and have a local instance serve as a git server, using your personal account for these tests is not optimal. 

The tests were not failing as they were never added to the main test suite, I cannot find a time when they were ever run on our side, I supposed that the case is that they were not working or would fail on the integration machines, but I don't remember this and might have not been involved at the time.

The issue you point at I closed as dupe as galaxy was mostly unimplemented in v2 (which is what I'm trying to finish up now) and I had several tickets pointing that out already, I did not look too closely at the tests at the time.

jimi-c commented:
# Closing This Ticket

Hi!

We believe the above commit should resolve this problem for you. This will also be included in the next major release.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!
